x-pack/legacy/plugins/apm/server/lib/service_map/run_service_map_task.ts

```diff
@@ -0,0 +1,248 @@
+
+  let mostRecent = '';
+  let afterKey = null;
+  while (true) {
+    const q = interestingTransactions(lastRun, afterKey);
+    const txs = await callCluster('search', {
+      index: apmIdxPattern,
+      body: q
+    });
+
+    if (txs.aggregations['ext-conns'].buckets.length < 1) {
+      return { mostRecent };
+    }
+    afterKey = txs.aggregations['ext-conns'].after_key;
+
+    const traces = new Set<string>();
+
+    txs.aggregations['ext-conns'].buckets.forEach((bucket: any) => {
+      bucket.smpl.tracesample.hits.hits.forEach((hit: any) => {
+        traces.add(hit._source.trace.id);
+        mostRecent =
+          mostRecent > hit._source['@timestamp']
+            ? mostRecent
+            : hit._source['@timestamp'];
+      });
+    });
+
+    const traceIds = Array.from(traces.values());
+    if (traceIds.length < 1) {
+      return { mostRecent: null };
+    }
+
+    const findConnsQ = findConns(traceIds);
+
+    const connsResult = await callCluster('search', {
+      index: apmIdxPattern,
+      body: findConnsQ
+    });
+
+    const connDocs: Array<{ index: { _index: any } } | ConnectionDoc> = [];
+
+    connsResult.aggregations.trace_id.buckets.forEach((bucket: any) => {
+      const allServices = new Set<string>(
+        bucket.connections.value.map(
+          (conn: any) =>
+            `${conn.caller.service_name}/${conn.caller.environment || 'null'}`
+        )
+      );
+
+      bucket.connections.value.forEach((conn: any) => {
+        const index = serviceConnsDestinationIndex
+          ? serviceConnsDestinationIndex
+          : conn.caller._index;
+        const bulkOpts = { index: { _index: index }, pipeline: undefined };
+
+        if (serviceConnsDestinationPipeline) {
+          bulkOpts.pipeline = serviceConnsDestinationPipeline;
+        }
+        connDocs.push(bulkOpts);
+        const doc: ConnectionDoc = {
+          '@timestamp': conn.caller.timestamp,
+          observer: { version_major: 7 }, // TODO: make this dynamic
+          service: {
+            name: conn.caller.service_name,
+            environment: conn.caller.environment
+          },
+          callee: conn.callee
+            ? {
+                name: conn.callee.service_name,
+                environment: conn.callee.environment
+              }
+            : undefined,
+          connection: {
+            upstream: {
+              list: conn.upstream,
+              keyword: conn.upstream.join('->')
+            },
+            in_trace: Array.from(allServices),
+            type: conn.caller.span_type,
+            subtype: conn.caller.span_substype
+          },
+          destination: conn.caller.destination
+            ? { address: conn.caller.destination }
+            : undefined
+        };
+
+        connDocs.push(doc);
+      });
+    });
+
+    const body = connDocs
+      .map((connDoc: any) => JSON.stringify(connDoc))
+      .join('\n');
+    await callCluster('bulk', {
+      body
+    });
+  }
+}
```

Comment:
As we talked about, I think this file would benefit from being refactored into smaller functions, having better types, and being less mutable (if possible).